test/e2e/storage/drivers/csi.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+ * This file defines various csi volume test drivers for TestSuites.
+ *
+ * There are two ways, how to prepare test drivers:
+ * 1) With containerized server (NFS, Ceph, Gluster, iSCSI, ...)
+ * It creates a server pod which defines one volume for the tests.
+ * These tests work only when privileged containers are allowed, exporting
+ * various filesystems (NFS, GlusterFS, ...) usually needs some mounting or
+ * other privileged magic in the server pod.
+ *
+ * Note that the server containers are for testing purposes only and should not
+ * be used in production.
+ *
+ * 2) With server or cloud provider outside of Kubernetes (Cinder, GCE, AWS, Azure, ...)
+ * Appropriate server or cloud provider must exist somewhere outside
+ * the tested Kubernetes cluster. CreateVolume will create a new volume to be
+ * used in the TestSuites for inlineVolume or DynamicPV tests.
+ */
+
+package drivers
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+// hostpathCSI
+type hostpathCSIDriver struct {
+	cleanup    func()
+	driverInfo DriverInfo
+}
+
+var _ TestDriver = &hostpathCSIDriver{}
+var _ DynamicPVTestDriver = &hostpathCSIDriver{}
+
+// InitHostPathCSIDriver returns hostpathCSIDriver that implements TestDriver interface
+func InitHostPathCSIDriver() TestDriver {
+	return &hostpathCSIDriver{
+		driverInfo: DriverInfo{
+			Name:        "csi-hostpath",
+			FeatureTag:  "",
+			MaxFileSize: testpatterns.FileSizeMedium,
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+			IsPersistent:       true,
+			IsFsGroupSupported: false,
+			IsBlockSupported:   false,
```

Comment:
@mkimuram it's not obvious what "IsPersistent" means, and the description ("Flag to represent whether it provides persistency") doesn't explain it either (it just paraphrases the variable name, which usually is a comment that might as well be left out because it doesn't add additional information).

Does it mean "persistent across node restarts"? The hostpath driver isn't that, so I'm wondering why it is set to true here.

I also don't know what "fsGroup" is.

"Block support" is a bit easier to guess, but an explicit description would be better.

Reply:
@pohly 

> it's not obvious what "IsPersistent" means

Actually, I meant "persistent across pod restarts".  It is defined to handle [emptyDir driver](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L899
) properly in [volumes test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/volumes.go#L72). I agree that we should improve the comment and we might need something like ```IsPersistentAcrossNode``` flag in the future.

> I also don't know what "fsGroup" is

Please see below.
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod

Most of these flags are defined in the process of migrating existing in-tree driver's test to this framework. So, we will need more flags in the future and improve these flags and their comments.

Reply:
I'm changing the way these capabilities are defined and hopefully making the comments more clear here: https://github.com/kubernetes/kubernetes/pull/70756

Reply:
@msau42 

Thank you for start working on improving capability definition.
The framework is ready for both in-tree and csi now, so it would be good timing to do it, before adding more testsuites and drivers.